WMF Framework/Remote Notifications/RemoteNotificationsOperationsController.swift

```diff
@@ -1,349 +1,250 @@
     }
 
     deinit {
         NotificationCenter.default.removeObserver(self)
     }
-
-    public func stop() {
-        operationQueue.cancelAllOperations()
-    }
     
-    /// Kicks off operations to fetch and persist read and unread history of notifications from app languages, Commons, and Wikidata. Designed to fully import once per installation. Will not attempt if import is already in progress. Must be called from main thread.
+    //MARK: Public
+    
+    /// Kicks off operations to fetch and persist read and unread history of notifications from app languages, Commons, and Wikidata, + other projects with unread notifications. Designed to automatically page and fully import once per installation, then only fetch new notifications for each project when called after that. Will not attempt if loading is already in progress. Must be called from main thread.
     /// - Parameter completion: Block to run once operations have completed. Dispatched to main thread.
-    func importNotificationsIfNeeded(_ completion: @escaping (RemoteNotificationsOperationsError?) -> Void) {
-        
+    func loadNotifications(_ completion: ((Result<Void, Error>) -> Void)? = nil) {
         assert(Thread.isMainThread)
         
-        guard !isLocked else {
-            assertionFailure("Failure setting up notifications core data stack.")
-            completion(.dataUnavailable)
-            return
+        if let completion = completion {
+            loadingNotificationsCompletionBlocks.append(completion)
         }
         
-        importingCompletionBlocks.append(completion)
-        
-        //Purposefully not calling completion block here, because we are tracking it in line above. It will be called when
-        //currently running operation completes.
-        guard !isImporting else {
+        //Purposefully not calling completion block here, because we are tracking it in line above. It will be called when currently running loading operations complete.
+        guard !isLoadingNotifications else {
             return
         }
         
-        isImporting = true
+        isLoadingNotifications = true
+        NotificationCenter.default.post(name: Notification.Name.NotificationsCenterLoadingDidStart, object: nil)
         
-        kickoffPagingOperations(operationType: RemoteNotificationsImportOperation.self) { [weak self] error in
+        kickoffPagingOperations() { [weak self] result in
             DispatchQueue.main.async {
-                self?.isImporting = false
-                self?.importingCompletionBlocks.forEach { completionBlock in
-                    completionBlock(error)
+                self?.isLoadingNotifications = false
+                self?.loadingNotificationsCompletionBlocks.forEach { completionBlock in
+                    completionBlock(result)
                 }
 
-                self?.importingCompletionBlocks.removeAll()
+                self?.loadingNotificationsCompletionBlocks.removeAll()
```

Comment:
In practice the main thread assertions and serial queue operations may result in it all being fine, but this array mutation gives me a bit of thread safety anxiety.

Reply:
Yeah, that's fair. I do think it's safe but it would be nice to avoid keeping track of these completion blocks at all. The only reason I'm tracking them this way is if importing/refreshing takes a long time - I want to be able to react when it completes if it's already in the middle of it.

What do you think about removing the completion block param from this method entirely, and we have callers react to it via the `NotificationsCenterLoadingDidEnd` notification that posts? Originally I thought of `NotificationsCenterLoadingDidEnd` as only being used for a basic UI loading state, but I can explore using it for whatever the completion blocks are doing.

Reply:
Interesting! So essentially the idea is `RemoteNotificationsController` becomes an observer for the `LoadingDidEnd` notifications, it calls the `RemoteNotificationsOperationsController`'s `loadNotifications`, then upon receiving that notification inspects something (maybe `userInfo` on the notification) to know if it was a success or failure to then perform whatever actions it wants?

If your intent for those notifications was more to serve the UI though, maybe it's best not to overload them and stick with the blocks. This would help prevent its specific intent leaking into and influencing other parts of the app. I personally don't have a problem with the completion block approach – I think it actually nicely draws a pretty clear path of intent for the reader, which is helpful when tracing and debugging.

Reply:
Yep! That's what I was thinking, but if the intent seems clear to you as it is then maybe I'll hold off until we're sure it'll be a problem.